src/shared/varlink.c

```diff
@@ -462,7 +481,44 @@ static int varlink_write(Varlink *v) {
          * Use a local variable to help gcc figure out that we set 'n' in all cases. */
         bool prefer_write = v->prefer_read_write;
         if (!prefer_write) {
-                n = send(v->fd, v->output_buffer + v->output_buffer_index, v->output_buffer_size, MSG_DONTWAIT|MSG_NOSIGNAL);
+                struct msghdr msgh = {0};
+                struct iovec iov;
+                union {
+                        char buf[CMSG_SPACE(sizeof(struct ucred))];
+                        struct cmsghdr align;
+                } cmsg;
+                struct cmsghdr *cmsgp;
+                struct ucred *creds = NULL;
+
+                iov.iov_base = v->output_buffer + v->output_buffer_index;
+                iov.iov_len = v->output_buffer_size;
+                msgh.msg_iov = &iov;
+                msgh.msg_iovlen = 1;
+
+                if (v->ucred_passed) {
+                        creds = &v->ucred_effective;
+                        log_debug("Sending previously passed creds for uid %d", (int)v->ucred_effective.uid);
+                } else {
+                        varlink_acquire_ucred(v);
+                        if (v->ucred_acquired) {
+                                creds = &v->ucred;
```

Comment:
hmm, SO_PEERCRED will acquire the peer's ucred, which you then use here for *sending* so the other side will see messages coming in from itself? What's the reason for that?